File: sdk/python/soluna/player.py
"""
Optional audio playback using sounddevice.
Requires: pip install sounddevice
"""

from __future__ import annotations

import array
import struct
from typing import TYPE_CHECKING

from .adpcm import decode_adpcm_payload
from .parser import OSTConstants, OSTPacket

if TYPE_CHECKING:
    import numpy as np

SAMPLE_RATE = 48000


class AudioPlayer:
    """Plays decoded PCM audio through sounddevice (PortAudio).
    Format: 48kHz stereo float32.
    """

    def __init__(self) -> None:
        self._stream = None
        self._sd = None

    def start(self) -> None:
        """Start the audio output stream."""
        try:
            import sounddevice as sd  # type: ignore
        except ImportError:
            raise ImportError("pip install sounddevice")

        self._sd = sd
        self._stream = sd.OutputStream(
            samplerate=SAMPLE_RATE,
            channels=2,
            dtype="float32",
            blocksize=0,
        )
        self._stream.start()

    def stop(self) -> None:
        """Stop the audio output stream."""
        if self._stream:
            self._stream.stop()
            self._stream.close()
            self._stream = None

    @property
    def is_playing(self) -> bool:
        return self._stream is not None and self._stream.active

    def play_packet(self, packet: OSTPacket) -> None:
        """Decode and play an OSTP packet."""
        if not self._stream:
            return

        import numpy as np  # type: ignore

        if packet.payload_type in (OSTConstants.PT_ADPCM_STEREO, OSTConstants.PT_ADPCM_MONO):
            samples = self._decode_adpcm_to_float(packet.payload, packet.channels)
        elif packet.payload_type == OSTConstants.PT_OPUS:
            samples = self._decode_opus_to_float(packet.payload, packet.channels)
        elif packet.payload_type == OSTConstants.PT_LC3:
            samples = self._decode_lc3_to_float(packet.payload, packet.channels)
        else:
            samples = self._decode_int32le_to_float(packet.payload, packet.channels)

        if samples is None or len(samples) == 0:
            return

        channels = min(packet.channels, 2)
        frames = len(samples) // channels

        if frames == 0:
            return

        # Reshape to (frames, channels)
        interleaved = np.array(samples[:frames * channels], dtype=np.float32)
        if channels == 1:
            # Mono -> stereo
            mono = interleaved.reshape(-1, 1)
            stereo = np.column_stack([mono, mono])
        else:
            stereo = interleaved.reshape(-1, channels)[:, :2]

        try:
            self._stream.write(stereo)
        except Exception:
            pass

    def _decode_int32le_to_float(self, payload: bytes, channels: int) -> list[float] | None:
        """Decode int32 LE interleaved payload to float32 samples."""
        total_samples = len(payload) // 4
        if total_samples == 0:
            return None

        scale = 1.0 / 2147483647.0
        samples = []
        for i in range(total_samples):
            offset = i * 4
            val = struct.unpack_from("<i", payload, offset)[0]
            samples.append(val * scale)
        return samples

    def _decode_opus_to_float(self, payload: bytes, channels: int) -> list[float] | None:
        """Decode Opus payload to float32 samples.

        Requires: pip install opuslib
        """
        try:
            import opuslib  # type: ignore
        except ImportError:
            print("[SolunaSDK] Opus codec requires opuslib — pip install opuslib")
            return None

        try:
            decoder = opuslib.Decoder(48000, channels)
            # 960 frames = 20ms at 48kHz
            pcm = decoder.decode(payload, 960)
            # opuslib returns bytes (int16 LE), convert to float32
            total_samples = len(pcm) // 2
            samples = []
            for i in range(total_samples):
                val = struct.unpack_from("<h", pcm, i * 2)[0]
                samples.append(val / 32768.0)
            return samples
        except Exception as e:
            print(f"[SolunaSDK] Opus decode error: {e}")
            return None

    def _decode_lc3_to_float(self, payload: bytes, channels: int) -> list[float] | None:
        """Decode LC3 payload to float32 samples.

        LC3 decode requires liblc3 (Google, Apache 2.0).
        https://github.com/google/liblc3
        TODO: Add Python bindings for liblc3.
        """
        print("[SolunaSDK] LC3 codec not yet supported — install liblc3")
        return None

    def _decode_adpcm_to_float(self, payload: bytes, channels: int) -> list[float] | None:
        """Decode ADPCM payload to float32 samples."""
        pcm16 = decode_adpcm_payload(payload)
        if pcm16 is None:
            return None

        return [s / 32768.0 for s in pcm16]

File: sdk/rust/src/player.rs
//! Optional audio playback using cpal.
//! Enable with the `audio` feature.

use crate::adpcm::decode_adpcm_payload;
use crate::parser::{constants, OSTPacket};

/// Decode a packet's payload to interleaved float32 PCM samples.
pub fn decode_packet_to_f32(packet: &OSTPacket) -> Option<Vec<f32>> {
    if packet.payload_type == constants::PT_ADPCM_STEREO
        || packet.payload_type == constants::PT_ADPCM_MONO
    {
        decode_adpcm_to_f32(&packet.payload)
    } else if packet.payload_type == constants::PT_OPUS {
        decode_opus_to_f32(&packet.payload, packet.channels)
    } else if packet.payload_type == constants::PT_LC3 {
        decode_lc3_to_f32(&packet.payload, packet.channels)
    } else {
        decode_int32le_to_f32(&packet.payload)
    }
}

/// Decode int32 LE interleaved payload to float32.
fn decode_int32le_to_f32(payload: &[u8]) -> Option<Vec<f32>> {
    let total_samples = payload.len() / 4;
    if total_samples == 0 {
        return None;
    }

    let scale = 1.0 / i32::MAX as f32;
    let mut out = Vec::with_capacity(total_samples);

    for i in 0..total_samples {
        let offset = i * 4;
        if offset + 3 >= payload.len() {
            break;
        }
        let val = i32::from_le_bytes([
            payload[offset],
            payload[offset + 1],
            payload[offset + 2],
            payload[offset + 3],
        ]);
        out.push(val as f32 * scale);
    }

    Some(out)
}

/// Decode Opus payload to float32.
///
/// Requires the `opus` feature (audiopus crate).
/// Returns None if the feature is not enabled or decode fails.
#[cfg(feature = "opus")]
fn decode_opus_to_f32(payload: &[u8], channels: usize) -> Option<Vec<f32>> {
    use audiopus::{coder::Decoder, Channels, SampleRate};

    let ch = match channels {
        1 => Channels::Mono,
        _ => Channels::Stereo,
    };

    let mut decoder = Decoder::new(SampleRate::Hz48000, ch).ok()?;
    // 960 frames = 20ms at 48kHz
    let max_samples = 960 * channels;
    let mut output = vec![0.0f32; max_samples];

    match decoder.decode_float(Some(payload), &mut output, false) {
        Ok(decoded_samples) => {
            output.truncate(decoded_samples * channels);
            Some(output)
        }
        Err(e) => {
            eprintln!("[SolunaSDK] Opus decode error: {e}");
            None
        }
    }
}

#[cfg(not(feature = "opus"))]
fn decode_opus_to_f32(_payload: &[u8], _channels: usize) -> Option<Vec<f32>> {
    eprintln!("[SolunaSDK] Opus codec requires the 'opus' feature — add audiopus to Cargo.toml");
    None
}

/// Decode LC3 payload to float32.
///
/// LC3 decode requires liblc3 (Google, Apache 2.0).
/// https://github.com/google/liblc3
/// TODO: Add liblc3-sys bindings.
fn decode_lc3_to_f32(_payload: &[u8], _channels: usize) -> Option<Vec<f32>> {
    eprintln!("[SolunaSDK] LC3 codec not yet supported — install liblc3");
    None
}

/// Decode ADPCM payload to float32.
fn decode_adpcm_to_f32(payload: &[u8]) -> Option<Vec<f32>> {
    let pcm16 = decode_adpcm_payload(payload)?;
    Some(pcm16.iter().map(|&s| s as f32 / 32768.0).collect())
}

/// Convert interleaved samples to stereo (mono -> duplicate, multi-ch -> take first 2).
pub fn to_stereo(samples: &[f32], channels: usize) -> Vec<f32> {
    if channels == 2 {
        return samples.to_vec();
    }

    let frames = samples.len() / channels.max(1);
    let mut out = Vec::with_capacity(frames * 2);

    if channels == 1 {
        for &s in samples {
            out.push(s);
            out.push(s);
        }
    } else {
        for f in 0..frames {
            let base = f * channels;
            out.push(samples.get(base).copied().unwrap_or(0.0));
            out.push(samples.get(base + 1).copied().unwrap_or(0.0));
        }
    }

    out
}

#[cfg(feature = "audio")]
pub mod cpal_player {
    //! cpal-based audio output at 48kHz stereo float32.

    use cpal::traits::{DeviceTrait, HostTrait, StreamTrait};
    use std::sync::{Arc, Mutex};
    use std::collections::VecDeque;

    const SAMPLE_RATE: u32 = 48000;

    pub struct AudioPlayer {
        _stream: Option<cpal::Stream>,
        buffer: Arc<Mutex<VecDeque<f32>>>,
    }

    impl AudioPlayer {
        pub fn new() -> Self {
            Self {
                _stream: None,
                buffer: Arc::new(Mutex::new(VecDeque::new())),
            }
        }

        pub fn start(&mut self) -> Result<(), Box<dyn std::error::Error>> {
            let host = cpal::default_host();
            let device = host
                .default_output_device()
                .ok_or("No output device")?;

            let config = cpal::StreamConfig {
                channels: 2,
                sample_rate: cpal::SampleRate(SAMPLE_RATE),
                buffer_size: cpal::BufferSize::Default,
            };

            let buf = self.buffer.clone();
            let stream = device.build_output_stream(
                &config,
                move |data: &mut [f32], _info| {
                    let mut buf = buf.lock().unwrap();
                    for sample in data.iter_mut() {
                        *sample = buf.pop_front().unwrap_or(0.0);
                    }
                },
                |err| eprintln!("[SolunaSDK] cpal error: {err}"),
                None,
            )?;

            stream.play()?;
            self._stream = Some(stream);
            Ok(())
        }

        pub fn stop(&mut self) {
            self._stream = None;
        }

        /// Push interleaved stereo float32 samples.
        pub fn push_samples(&self, samples: &[f32]) {
            if let Ok(mut buf) = self.buffer.lock() {
                buf.extend(samples);
                // Cap buffer at ~500ms to prevent memory growth
                const MAX_SAMPLES: usize = 48000 * 2; // 500ms stereo
                while buf.len() > MAX_SAMPLES {
                    buf.pop_front();
                }
            }
        }
    }
}
